macros/compare_table_values.sql

```diff
@@ -0,0 +1,56 @@
+/*
+The [dbt Audit Helper package](https://github.com/dbt-labs/dbt-audit-helper#compare_column_values-source)
+includes a macro called `compare_column_values` that takes two queries and compares the values in a specific column.
+In the documentation, it gives an example of how to loop through all the columns in the two tables, _however_,
+this pattern does work because it relies on `agate.Table.print_table`, which outputs to `stdout`,
+which in turn is not supported in dbt Cloud (logs are sent to `stderr`).
+
+This macro wraps the `compare_column_values` macro to
+(1) Detect whether the dbt environment is dbt-core or dbt Cloud and output results to the log accordingly.
+(2) If executed in dbt Cloud, requires `dbt run-operation` to output the results to the logs.
+(3) Assumes the primary use case is comparing an old table in production to a modified version of the same table in the user's Develop environment.
+
+arguments:
+old_table: includes full namespace e.g. my_production_database.my_production_schema.my_production_table
+new_table_ref: model name in your Develop environment e.g. my_model_name -- not a ref object, just the name
+primary_key: must be unique in both for join. See https://github.com/dbt-labs/dbt-audit-helper#compare_column_values-source
+
+usage:
+dbt run-operation compare_table_values --args '{"old_table": "my_production_database.my_production_schema.my_production_table", "new_table_ref": "my_production_table", "primary_key": "my_primary_key"}'
+*/
+
+{% macro compare_table_values(old_table, new_table_ref, primary_key) %}
+{%- set columns_to_compare=adapter.get_columns_in_relation(ref(new_table_ref))  -%}
+
+{% set old_etl_relation_query %}
+    select * from {{ old_table }}
+{% endset %}
+
+{% set new_etl_relation_query %}
+    select * from {{ ref(new_table_ref) }}
+{% endset %}
+
+{% if execute %}
+    {% for column in columns_to_compare %}
+        {{ log('Comparing column "' ~ column.name ~'"', info=True) }}
+        {% set audit_query = audit_helper.compare_column_values(
+                a_query=old_etl_relation_query,
+                b_query=new_etl_relation_query,
+                primary_key=primary_key,
+                column_to_compare=column.name
+        ) %}
+
+        {% set audit_results = run_query(audit_query) %}
+        {% if env_var("DBT_ENV") == 'core' %}
+            {% do audit_results.print_table() %}
+            {{ log("", info=True) }}
+        {% elif env_var("DBT_ENV") == 'prod' %}
+            {% do log(audit_results.column_names, info=True) %}
+            {% for row in audit_results.rows %}
+                {% do log(row.values(), info=True) %}
+            {% endfor %}
+        {% endif %}
+    {% endfor %}
+{% endif %}
+
+{% endmacro %}
```

Comment:
When I ran this in my local environment, I got the following error because `DBT_ENV` wasn't set:
```
Encountered an error while running operation: Parsing Error
  Env var required but not provided: 'DBT_ENV'
```

A few different ideas of how to approach this:
1. Default to `'core'` if `DBT_ENV` is not set (i.e., `env_var("DBT_ENV", "core")`)
1. Document that `DBT_ENV` needs to be set to use in a non-Cloud environment
1. Skip the `DBT_ENV` check altogether do the same text output for both Cloud and non-Cloud

Here's what the first option could look like after creating a stand-alone macro to determine if Cloud or not and then flipping the order of the if/else blocks:

```suggestion
        {% set audit_results = run_query(audit_query) %}

        {% if is_cloud() %}
            {% do log(audit_results.column_names, info=True) %}
            {% for row in audit_results.rows %}
                {% do log(row.values(), info=True) %}
            {% endfor %}
        {% else %}
            {% do audit_results.print_table() %}
            {{ log("", info=True) }}
        {% endif %}
    {% endfor %}
{% endif %}

{% endmacro %}

{% macro is_cloud() %}
    {# In the dbt Cloud context, `DBT_ENV` will always resolve to "prod" #}
    {% set is_cloud = env_var("DBT_ENV", "core") == "prod" %}
    {{ return(is_cloud) }}
{% endmacro %}
```

Reply:
Thanks for reviewing @dbeatty10! Option 1 makes the most sense to me, compared to the additional manual setup work of 2. As for 3, I guess the questions I have end up being about how `dbt_env` is set. When I discussed this with dbt support back in July, I was told:

```
DBT_ENV: This key is reserved for the dbt Cloud application and will always resolve to 'prod'. So the person can call the . If it evaluates to prod, it's cloud, and if it evaluates to core it's core.
https://docs.getdbt.com/docs/dbt-cloud/using-dbt-cloud/cloud-environment-variables#special-environment-variables
```

Looking at the commit history for that page, I'm not sure that the page ever actually asserted that `dbt_env` resolves to "core" if running on core. Based on what you are saying, it sounds like on `dbt-core` the `dbt_env` by default is not set at all. If that is the case, it might be worth confirming and making explicit in the main dbt docs.

Coming back to option 3, I have not tested this on `dbt-core`, but it sounds like you did and the output looked correct? Here's what I get in Cloud. Note that it is not perfect but it gets the information out in the Cloud console.

```
13:49:31  Unable to do partial parsing because config vars, config profile, or config target have changed
13:50:12  Comparing column "ID"
13:50:14  (<agate.Row: ('✅: perfect match', Decimal('10788006'), Decimal('95.15'))>, <agate.Row: ('🤷: missing from b', Decimal('550162'), Decimal('4.85'))>)
```

Also, looking back at https://github.com/dbt-labs/dbt-audit-helper/pull/45, I believe the description of this PR mostly answers [the questions you had back in June](https://github.com/dbt-labs/dbt-audit-helper/pull/45#pullrequestreview-1021919675) about why `print_table()` does not work in Cloud, and how to introspect whether it's running on Cloud. Perhaps we use this PR to also amend the docs.

Reply:
### Bonus idea
One other idea to throw in the mix:
- What if we defined a `audit_helper.print_table()` macro? It would take an `agate.Table` as the only parameter. And then it would use your Cloud/non-Cloud logic to determine how best to print it.

The definition could look something like this:
```sql
{% macro print_table(audit_results) %}
{% if execute %}
    {% if is_cloud() %}
        {% do log(audit_results.column_names, info=True) %}
        {% for row in audit_results.rows %}
            {% do log(row.values(), info=True) %}
        {% endfor %}
    {% else %}
        {% do audit_results.print_table() %}
        {{ log("", info=True) }}
    {% endif %}
{% endif %}
{% endmacro %}
```

Then you could just use new `audit_helper.print_table` macro in the applicable place in your implementation. We could also replace any other references to `print_table()` with this macro instead. It looks like it only appears in the README and a couple integration tests: see [here](https://github.com/dbt-labs/dbt-audit-helper/search?q=print_table).

The big advantage is that both Cloud and non-Cloud users could both run all the same exact code examples and get results that they can view (albeit slightly different rendering of the table).

### One thing we didn't discuss

The output in a Core environment is something like this:
```
01:07:58  Comparing column "id"
| match_status     | count_records | percent_of_total |
| ---------------- | ------------- | ---------------- |
| ✅: perfect match |           100 |              100 |
```

Theoretically, this formatting could be mimic'd for the Cloud implementation. I'm fine with diverging formatting since this is meant to be human-readable rather than machine readable. If it were meant to be used by machines also, then we should definitely try to keep the formatting consistent. I'm also fine if you want to mimic the Agate `print_table` formatting.

### Next steps

> Option 1 makes the most sense to me, compared to the additional manual setup work of 2.

Awesome, let's go for it!

> Based on what you are saying, it sounds like on `dbt-core` the `dbt_env` by default is not set at all. If that is the case, it might be worth confirming and making explicit in the main dbt docs.

I can confirm that the `DBT_ENV` environment variable won't be set in Core environments by merely installing dbt-core. It would need to be explicitly set somehow. If we go with Option 1, any Core user that does something like `export DBT_ENV=prod` will get the Cloud-specific behavior of this macro. I would feel comfortable with that.

> Coming back to option 3, I have not tested this on `dbt-core`, but it sounds like you did and the output looked correct?

Yep, here's what I got when running the Cloud-specific code in a Core environment (which feels readable enough):
```
00:50:00  Running with dbt=1.3.1
00:50:01  Comparing column "id"
00:50:02  ('match_status', 'count_records', 'percent_of_total')
00:50:02  ('✅: perfect match', Decimal('100'), Decimal('100.0'))
00:50:02  Comparing column "first_name"
00:50:03  ('match_status', 'count_records', 'percent_of_total')
00:50:03  ('✅: perfect match', Decimal('100'), Decimal('100.0'))
00:50:03  Comparing column "last_name"
00:50:04  ('match_status', 'count_records', 'percent_of_total')
00:50:04  ('✅: perfect match', Decimal('100'), Decimal('100.0'))
```

> Also, looking back at #45, I believe the description of this PR mostly answers [the questions you had back in June](https://github.com/dbt-labs/dbt-audit-helper/pull/45#pullrequestreview-1021919675) about why `print_table()` does not work in Cloud, and how to introspect whether it's running on Cloud. Perhaps we use this PR to also amend the docs.

Yeah, let's amend the docs while we're at it 💪

Reply:
great! I'd love to get the formatting consistent but have not yet dove into it. Working on the other bits.